Hour of Code 2015 studio work

breville opened the pull request.

breville commented:
fyi @bcjordan @Bjvanminnen and @cpirich

Bjvanminnen commented on apps/src/studio/Item.js:
```diff
@@ -18,12 +18,15 @@ var Item = function (options) {
+    if (self.dir != Direction.NONE) {
+      if (self.loop || self.currentFrame_ + 1 < self.frames) {
+        self.currentFrame_ = (self.currentFrame_ + 1) % self.frames;
+      }
```
Could combine if statement into a single line. Alternatively, could invert and reducing nesting, i.e...
```
if (self.dir === Direction.NONE) {
  return;
}
if (self.loop...) {
  self.currentFrame_ = ...
}
```

Also, instead of using self, we could just bind the function inside of setInterval to this, i.e.

```
window.setInterval(function () {
  if (this.dir != Direciton.NONE) {
  ...
}.bind(this), 50)

Bjvanminnen replied:
Also, our linter will hopefully complain about != instead of !==

Bjvanminnen commented on apps/src/studio/Item.js:
```diff
@@ -46,7 +49,7 @@ Item.__resetIds = function () {
   // create our clipping path/rect
   this.clipPath = document.createElementNS(SVG_NS, 'clipPath');
-  var clipId = 'item_clippath_' + (uniqueId++);
+  var clipId = 'item_clippath_' + uniqueId;
```
What if we have `var nextId = (uniqueId++);` at the top of this function, and then use `nextId` instead of `uniqueId` throughout. Might be a little less confusing than mutating uniqueId somewhere in the middle.

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -2132,6 +2228,11 @@ var ANIM_AFTER_NUM_NORMAL_FRAMES = 8;
 var IDLE_TICKS_BEFORE_FACE_SOUTH = 4;
 
+Studio.itemGetDirectionFrame = function(item) {
+  return item.dir == Direction.NONE ? 0 : frameDirTableWalking[item.dir];
```
Might be better to have frameDirTableWalking live in and get exported from src/studio/constants.js, allowing us to have getDirectionFrame live in Item.js. We could also add an entry:
```
frameDirTableWalking[Direction.NONE]      = 0;
```

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
 };
 
+function getRandomInt(min, max) {
+    return Math.floor(Math.random() * (max - min + 1)) + min;
```
too much indent

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+  if (this.gridX === undefined) {
+    this.gridX = Math.floor(this.x / Studio.SQUARE_SIZE);
+    this.gridY = Math.floor(this.y / Studio.SQUARE_SIZE);
+  }
```
Something I like to do to make classes more readable is to take any fields that are going to exist on the object, and initialize them in the constructor. This way I can look at the ctor and get an idea of what all exists on the class. So in this case this would be something like
```
var Collidabale = function () {
  ...
  this.gridX = undefined;
  this.gridY = undefined;
}

brandonbloom replied:
Initializing all fields in the constructor is also more likely to trigger "hidden classes" optimizations in the JavaScript runtime!

bcjordan replied:
Ooh, neat!

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+    return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+Collidable.prototype.roamGrid = function(type) {
```
This looks to be a fairly long function. It would be great to have a short comment explaining what it does, as well as what the expectations for the param "type" are.

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+
+    var bufferDistance = 60;
+
+    for (var candidateX = this.gridX - 1; candidateX <= this.gridX + 1; candidateX++) {
```
Would it make sense to try extracting some of this logic into its own function?

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+      if (atEdge || hasWall || candidate.score === 0) {
+        candidates.splice(i, 1);
+      }
+    }
```
Couldn't we just do this logic before deciding to push onto our array instead of pushing everything, and then stripping some?

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+      // then sort everything based on score.
+      candidates.sort(function (a, b) {
+        return b.score - a.score;
+      });
```
Not sure how many candidates we expect to have, but it seems like it would be more efficient to sort, pull out the top candidates, and then only shuffle those (since it looks like we just want one of the top candidates). Could be that this is an unnecessary perf optimization.

Bjvanminnen commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +108,157 @@ Collidable.prototype.bounce = function () {
+        this.dir = Direction.NONE;
+      }
+    }
+    else {
```
else on same line as previous line

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -587,7 +595,7 @@ var setSvgText = function(opts) {
  *  be appended to the queue
  */
-function callHandler (name, allowQueueExtension) {
+function callHandler (name, allowQueueExtension, extraArgs) {
```
Add an @param tag for extraArgs

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -637,6 +645,51 @@ function performItemOrProjectileMoves (list) {
+
+  itemsArray.sort(function(a, b) {
+    return a.y - b.y;
+  });
```
could also do `itemsArray = _.sortBy(itemsArray, 'y');

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -637,6 +645,51 @@ function performItemOrProjectileMoves (list) {
+ * z-sorting.
+ */
+function sortDrawOrder() {
+  var list = document.getElementById('svgStudio');
```
Maybe clearer to call this svg instead of list?

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -789,8 +850,12 @@ function spriteCollisionDistance  (i1, i2, yAxis) {
 }
 
 function spriteCollidableCollisionDistance (iS, collidable, yAxis) {
```
Realize this is existing code, but while you're in here could you add @param comments for these? `iS` seems particularly confusing to me initially. Presumably this is meant to be spriteIndex?

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -991,7 +1068,7 @@ function createItemEdgeCollisionHandler (item) {
     if (level.blockMovingIntoWalls) {
       // TODO: Find a real direction!!
-      item.bounce();
+      //item.bounce();
```
intentionally commented out?

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -1002,6 +1079,13 @@ function createItemEdgeCollisionHandler (item) {
   };
 }
 
+function updateItems () {
```
Could use a short description of what we're updating about these items.

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -1011,7 +1095,7 @@ function checkForItemCollisions () {
         if (level.blockMovingIntoWalls) {
           // TODO: Find a real direction!!
-          item.bounce();
+          //item.bounce();
```
intentionally commented out?

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -2207,40 +2308,84 @@ function cellId(prefix, row, col) {
+  $(".roamGridDest").remove();
+  $(".itemCenter").remove();
+  $(".roamGridPossibleDest").remove();
+};
```
You could give each of your debug rects a unique class and a shared class (i.e. spriteCenter debugRect) which would then allow you to just do $(".debugRect").remove().
Is there a reason that each of these need unique classNames?

Bjvanminnen commented on apps/src/studio/studio.js:
```diff
@@ -2701,6 +2865,20 @@ Studio.addItemsToScene = function (opts) {
+        item.roamGrid(opts.type);
+    }
+  }
+};
```
Could combine conditionals into a single one, or just do
```
if (!item) {
  return;
}
```
to reduce nesting

Bjvanminnen commented:
Cool stuff. Looks like you've made a lot of progress :)

breville commented:
Thanks @Bjvanminnen .  I addressed all of the feedback apart from modifications to the roamGrid function, which I want to save up for future updates since I think it'll change the most shortly.

bcjordan commented on apps/src/studio/collidable.js:
```diff
@@ -104,6 +111,167 @@ Collidable.prototype.bounce = function () {
+    }
+
+    // cull candidates that won't be possible
+    for (var i = candidates.length-1; i >= 0; i--) {
```
Could alternatively use `_.remove(candidates, myShouldRemoveCandidateFn)` for this

bcjordan commented on apps/src/studio/studio.js:
```diff
@@ -772,28 +833,54 @@ Studio.onTick = function() {
 
+/**
+ * Returns the distance between a sprite and a collidable on the specified axis.
+ * @param {number} i1 The index of the sprite.
```
Looks like comment param names / descriptions could use updating

bcjordan commented on apps/src/studio/studio.js:
```diff
@@ -772,28 +833,54 @@ Studio.onTick = function() {
 
+/**
+ * Returns the distance between a collidable and an edge on the specified axis.
+ * @param {number} i1 The index of the collidable.
```
Ditto comment change

bcjordan commented on apps/src/studio/studio.js:
```diff
@@ -1002,6 +1097,15 @@ function createItemEdgeCollisionHandler (item) {
   };
 }
 
+/* Calls each item's update function (if/when provided by student).
```
+newline

bcjordan commented on apps/src/studio/studio.js:
```diff
@@ -2207,40 +2302,69 @@ function cellId(prefix, row, col) {
+/**
+ * Clear the debug rectangles.
+ */
+
```
extra \n

bcjordan commented:
@breville LGTM! Nothing but nits. Only thing that may become useful to extract down the line would be some grid helpers, e.g. `absoluteToGrid(x, y)`, `gridToAbsolute(gridX, gridY)`, `forEachGridSpace(myFunction)`
